File: BloodDonationSystem.java
package Object_Oriented_Programming;

import java.util.*;
import java.io.*;


// =========================================
// ABSTRACT INTERFACE (ABSTRACTION)
// =========================================

// ABSTRACTION: The interface hides HOW searching works.
// Only the method names are exposed—implementation happens elsewhere.
// This is a key example of ABSTRACTION, as it defines a contract without specifying details.
interface Searchable {
    void searchByName(String name);
    void searchByBloodType(String type);
}


// =========================================
// PARENT CLASS (INHERITANCE + ENCAPSULATION)
// =========================================
class Person {

    // ENCAPSULATION: Fields are private and accessed only through getters.
    // This demonstrates ENCAPSULATION by protecting data and controlling access.
    private String name;
    private int age;
    private String contact;

    // Constructor initializes encapsulated data.
    public Person(String name, int age, String contact) {
        this.name = name;
        this.age = age;
        this.contact = contact;
    }

    // Method to display common info for all persons.
    public void displayInfo() {
        System.out.println("Name: " + name);
        System.out.println("Age: " + age);
        System.out.println("Contact: " + contact);
    }

    // ENCAPSULATION: Controlled access to private fields via getters.
    public String getName() { return name; }
    public int getAge() { return age; }
    public String getContact() { return contact; }
}


// =========================================
// CHILD CLASS (INHERITANCE + POLYMORPHISM)
// =========================================

// INHERITANCE: Donor inherits fields/methods from Person.
// This shows INHERITANCE, allowing Donor to reuse Person's attributes and methods.
// POLYMORPHISM: Donor overrides displayInfo() to add more info.
// This demonstrates POLYMORPHISM through method overriding.
class Donor extends Person {

    // Additional attributes only a Donor has.
    private String bloodType;
    private double weight;

    // Inherits name, age, and contact from Person via super().
    // INHERITANCE: Using super() to call parent constructor.
    public Donor(String name, int age, String bloodType, double weight, String contact) {
        super(name, age, contact);
        this.bloodType = bloodType;
        this.weight = weight;
    }

    // POLYMORPHISM: Overriding the displayInfo() of the parent class.
    // This is POLYMORPHISM, as the method behaves differently for Donor objects.
    @Override
    public void displayInfo() {
        // Calls parent method → polymorphic behavior.
        super.displayInfo();
        System.out.println("Blood Type: " + bloodType);
        System.out.println("Weight: " + weight);
    }

    // Encapsulated getter for extra fields
    // ENCAPSULATION: Providing controlled access to private fields.
    public String getBloodType() { return bloodType; }
    public double getWeight() { return weight; }
}


// =========================================
// DONOR DATABASE
// =========================================
class DonorArray {

    // ENCAPSULATION: donor list is private and only modifiable through methods.
    // This ensures ENCAPSULATION by hiding the list and controlling modifications.
    private ArrayList<Donor> donors = new ArrayList<>();

    public void addDonor(Donor d) {
        donors.add(d);
    }

    // Returns list for searching
    public ArrayList<Donor> getList() {
        return donors;
    }
}


// =========================================
// SEARCH MODULE (POLYMORPHISM)
// =========================================

// POLYMORPHISM: Implements Searchable interface → must override methods.
// This shows POLYMORPHISM through interface implementation and method overriding.
class DonorSearch implements Searchable {

    // Composition: Uses donor list to search.
    private ArrayList<Donor> donors;

    public DonorSearch(ArrayList<Donor> donors) {
        this.donors = donors;
    }

    // POLYMORPHISM: Implementing interface-defined methods.
    // Overriding methods from Searchable interface demonstrates POLYMORPHISM.
    @Override
    public void searchByName(String name) {
        for (Donor d : donors) {
            if (d.getName().equalsIgnoreCase(name)) {
                d.displayInfo(); // Polymorphic call (overridden method)
                return;
            }
        }
        System.out.println("Donor not found.");
    }

    @Override
    public void searchByBloodType(String type) {
        boolean found = false;
        for (Donor d : donors) {
            if (d.getBloodType().equalsIgnoreCase(type)) {
                d.displayInfo(); // Polymorphic call
                found = true;
            }
        }
        if (!found) {
            System.out.println("No donors match that blood type.");
        }
    }
}


// =========================================
// BLOOD INVENTORY SYSTEM
// =========================================
class BloodInventory {

    // ENCAPSULATION: inventory hidden and modified only through class methods.
    // This demonstrates ENCAPSULATION by keeping the map private and using methods for access.
    private HashMap<String, Integer> inventory = new HashMap<>();

    public BloodInventory() {
        inventory.put("A+",0); inventory.put("A-",0);
        inventory.put("B+",0); inventory.put("B-",0);
        inventory.put("O+",0); inventory.put("O-",0);
        inventory.put("AB+",0); inventory.put("AB-",0);
    }

    public void addBlood(String type, int units) {
        inventory.put(type, inventory.getOrDefault(type, 0) + units);
        System.out.println(units + " units added.");
    }

    public void viewInventory() {
        System.out.println("\n=== Available Blood Units ===");
        for (String type : inventory.keySet()) {
            System.out.println(type + ": " + inventory.get(type));
        }
    }
}


// =========================================
// MAIN SYSTEM (USES POLYMORPHISM FOR SEARCH)
// =========================================
public class BloodDonationSystem {

    private static Scanner sc = new Scanner(System.in);
    private static BloodInventory inventory = new BloodInventory();
    private static DonorArray donorDB = new DonorArray();

    public static void main(String[] args) {

        int choice;

        do {
            System.out.println("\n=== BLOOD DONATION SYSTEM ===");
            System.out.println("1. Register Donor");
            System.out.println("2. Check Eligibility");
            System.out.println("3. Blood Inventory");
            System.out.println("4. Search Donor");
            System.out.println("5. Exit");
            System.out.print("Enter choice: ");

            choice = sc.nextInt();
            sc.nextLine();

            switch (choice) {
                case 1: registerDonor(); break;
                case 2: eligibilityChecker(); break;
                case 3: inventoryMenu(); break;
                case 4: searchDonor(); break;
                case 5: System.out.println("Thank you for using the system!"); break;
                default: System.out.println("Invalid choice.");
            }

        } while (choice != 5);
    }


    // ___________________________
    // REGISTER DONOR
    private static void registerDonor() {

        System.out.println("\n== Donor Registration ==");

        System.out.print("Name: ");
        String name = sc.nextLine();

        System.out.print("Age: ");
        int age = sc.nextInt();

        System.out.print("Weight: ");
        double weight = sc.nextDouble();
        sc.nextLine();

        System.out.print("Blood Type (O, A, B, AB): ");
        String blood = sc.nextLine().toUpperCase();

        System.out.print("Rhesus (+ or -): ");
        String rh = sc.nextLine();

        System.out.print("Contact: ");
        String contact = sc.nextLine();

        Donor donor = new Donor(name, age, blood + rh, weight, contact);
        donorDB.addDonor(donor);

        System.out.println("Donor registered successfully!");
    }


    // ___________________________
    // ELIGIBILITY CHECKER
    private static void eligibilityChecker() {
        System.out.println("\n== Eligibility Checker ==");

        System.out.print("Age: ");
        int age = sc.nextInt();

        System.out.print("Weight: ");
        double weight = sc.nextDouble();

        System.out.print("Months since last donation: ");
        int months = sc.nextInt();
        sc.nextLine();

        System.out.print("Any illness? (yes/no): ");
        String ill = sc.nextLine();

        boolean eligible = (age >= 18 && age <= 65) &&
                           (weight >= 50) &&
                           (months >= 3) &&
                           (ill.equalsIgnoreCase("no"));

        if (eligible)
            System.out.println("You are eligible to donate!");
        else
            System.out.println("You are NOT eligible to donate.");
    }


    // ___________________________
    // INVENTORY MENU
    private static void inventoryMenu() {
        int choice;

        do {
            System.out.println("\n== Blood Inventory ==");
            System.out.println("1. Add Blood Units");
            System.out.println("2. View Inventory");
            System.out.println("3. Back");
            System.out.print("Enter: ");

            choice = sc.nextInt();
            sc.nextLine();

            switch (choice) {
                case 1:
                    System.out.print("Blood Type: ");
                    String type = sc.nextLine().toUpperCase();

                    System.out.print("Units: ");
                    int units = sc.nextInt();

                    inventory.addBlood(type, units);
                    break;

                case 2:
                    inventory.viewInventory();
                    break;
            }
        } while (choice != 3);
    }


    // ___________________________
    // SEARCH DONOR
    private static void searchDonor() {
        System.out.println("\n== Search Donor ==");
        System.out.println("1. By Name");
        System.out.println("2. By Blood Type");
        System.out.print("Enter: ");

        int opt = sc.nextInt();
        sc.nextLine();

        // POLYMORPHISM: Searchable reference pointing to DonorSearch object
        // This demonstrates POLYMORPHISM, as Searchable can refer to any implementing class.
        Searchable search = new DonorSearch(donorDB.getList());

        if (opt == 1) {
            System.out.print("Enter name: ");
            search.searchByName(sc.nextLine());
        }
        else if (opt == 2) {
            System.out.print("Enter blood type: ");
            search.searchByBloodType(sc.nextLine());
        }
    }
}

File: BloodDonationSystem.py
import mysql.connector
from mysql.connector import Error
from abc import ABC, abstractmethod


"""ABSTRACTION: The Searchable class is an abstract base class that defines the interface for searching by name and blood type. It abstracts away the details of how the search is implemented, allowing different classes to provide their own specific implementations while adhering to a common interface. This promotes a clear separation of concerns and allows for flexibility in how searching is performed across different classes."""

# Abstract class for the blood donation system

class Searchable(ABC):

    @abstractmethod
    def search_by_name(self, name):
        """Search for a record by name"""
        pass

    @abstractmethod
    def search_by_blood_type(self, blood_type):
        """Search for a record by blood type"""
        pass


"""INHERITANCE: The Searchable class can be inherited by other classes (like Donor or Recipient) to implement specific search functionalities. This promotes code reuse and establishes a natural hierarchical relationship between classes."""

class Person:
    def __init__(self, name: str, age: int, contact: str):
        self.__name = name
        self.__age = age
        self.__contact = contact

        """ENCAPSULATION: The attributes are private (using __) and can only be accessed through methods. This hides the internal state and protects it from direct access, ensuring controlled interaction with the data."""

    def display_info(self):
        print(f"Name: {self.__name}")
        print(f"Age: {self.__age}")
        print(f"Contact: {self.__contact}")

    def get_name(self):
        return self.__name
    
    def get_age(self):
        return self.__age
    
    def get_contact(self):
        return self.__contact
    
class Donor(Person):

    """INHERITANCE: The Donor class inherits from the Person class, allowing it to reuse the attributes and methods of Person while adding its own specific attributes (blood type and weight) and behaviors (displaying donor-specific information)."""

    def __init__(self, name: str, age: int, contact: str, blood_type: str, weight: float):
        super().__init__(name, age, contact)
        self.__blood_type = blood_type
        self.__weight = weight

    def display_info(self):
        super().display_info()
        print(f"Blood Type: {self.__blood_type}")
        print(f"Weight: {self.__weight} kg")

    def get_blood_type(self):
        return self.__blood_type
    
    def get_weight(self):
        return self.__weight
    
    """POLYMORPHISM: The display_info method is overridden in the Donor class to provide specific details about donors. This allows the same method name to behave differently based on the context (i.e., whether it's called on a Person or a Donor), demonstrating polymorphism."""

class BloodDonationDatabase(Searchable):

    """
    Will be using MariaDB to store and manage donor and recipient information. This class implements the Searchable interface, providing concrete implementations for searching by name and blood type.
    """

    def __init__(self, host, user, password, database):
        self.__config = {'host': host, 'user': user, 'password': password, 'database': database}

    def __get_connection(self):
        return mysql.connector.connect(**self.__config)
    
    def search_by_name(self, name: str):
        query = "SELECT name, age, contact, blood_type, weight FROM donors WHERE name = %s"

        try:
            connection = self.__get_connection()
            cursor = connection.cursor()

            cursor.execute(query, (name,))
            row = cursor.fetchone()

            if row:
                return Donor(name=row[0], age=row[1], contact=row[2], blood_type=row[3], weight=row[4])
            
            return None # return None of there are no results or does not exist
        
        except Error as e:
            print(f"Error while connecting to database: {e}")
            return None
        finally:
            if connection.is_connected():
                cursor.close()
                connection.close()

    def search_by_blood_type(self, blood_type: str):
        """Queries the database and returns a list of Donor objects matching the blood type."""
        query = "SELECT name, age, contact, blood_type, weight FROM donors WHERE blood_type = %s"
        donor_list = [] # Initialize an empty list to hold our results

        try:
            connection = self.__get_connection()
            cursor = connection.cursor()
            cursor.execute(query, (blood_type,))
            
            # 1. Fetch ALL rows that match (e.g., all 'O+' donors)
            rows = cursor.fetchall() 

            # 2. Loop through every row found
            for row in rows:
                # Create a Donor object for the current row
                matching_donor = Donor(
                    name=row[0], 
                    age=row[1], 
                    contact=row[2], 
                    blood_type=row[3], 
                    weight=row[4]
                )
                # Add this donor to our list
                donor_list.append(matching_donor)
            
            # 3. Return the populated list (will be empty [] if no matches found)
            return donor_list
        
        except Error as e:
            print(f"Error while connecting to database: {e}")
            return [] # Return an empty list if a database error happens
        finally:
            if connection.is_connected():
                cursor.close()
                connection.close()